google-cloud-storage/src/main/java/com/google/cloud/storage/StorageImpl.java

```diff
@@ -204,6 +208,24 @@ public StorageObject call() {
     }
   }
 
+  @Override
+  public void upload(BlobInfo blobInfo, Path path, BlobWriteOption... options) throws IOException {
```

Comment:
If these could be static you don't need to break the API.

Reply:
To upload data we need to get either a WriterChannel instance of a Storage instance.
To not brake the API we can add:
void Blob.upload(path) // this requires a Blob instance to be created before
or/and
static void Blob.upload(storage, path)

Perhaps this adding just static Blob.upload(storage, path/input, options) will be a compromise between the convenient  API (void Storage.upload(path/input)  and incompatible change.

Reply:
I think you're absolutely right, adding helper methods should not break compatibility.
Would it be okay to add a class like:
```
public final class StorageUtils {
    public final Storage storage;
   
    private StorageUtils(Storage storage) {
        this.storage = storage;
    }
   
    public static StorageUtils getInstance(Storage storage) {
        return new StorageUtils(storage);
    }

    public void upload(BlobInfo blobInfo, Path path, Storage.BlobWriteOption... options) throws IOException {
    }

    public void upload(BlobInfo blobInfo, InputStream content, Storage.BlobWriteOption... options) throws IOException {
    }

    static void upload(InputStream input, WriteChannel writer) throws IOException {
    }

    public static void upload(InputStream input, WriteChannel writer, int bufferSize) throws IOException  {
    }
}
```
- appropriate place 
- no API break
- Blob instance is not required
- new features could implemented here: folder upload, parallel_composite_upload,

Reply:
It's certainly OK from an API compatibility perspective. You can almost always add new non-abstract methods to classes, just not to interfaces.

I do find the mix of channels and input streams in a single method call surprising. Most APIs use old I/o (streams) or new I/O (channels) but not both at the same time. Not sure it's bad, but it's worth a close look. 

Honestly, channels and nio never really lived up to the hype. They were too late to the party. By the time they arrived, operating systems had improved to the point where running a bunch of threads with streams ran as fast or faster.

Reply:
This is a good point about the mix, I never thought this aspect. 
Probably instead of the mix it would be more elegant to introduce:
`upload(ReadableByteChannel reader, WritableByteChannel writer, int bufferSize)`

And provide the example in the comment how to obtain instances of reader/writer from file/storage.